### PR TITLE
[CCFPCM-419] pos reconciliation service and pos deposit service unit tests

### PR DESCRIPTION
**Pull request by chelsea-EYDS**
[CCFPCM-0419](https://bcdevex.atlassian.net/browse/CCFPCM-0419)

Objective:

- POS reconciliation service unit tests
- POS deposit unit tests

### CODE REVIEW

File: apps/backend/test/unit/reconciliation/pos-reconciliation.spec.ts [hunk 1]
```diff
@@ -1,58 +1,311 @@
...
+    const mockPOSData = new MockData(PaymentMethodClassification.POS);
+
+    payments = mockPOSData.paymentsMock;
+    /**
+     * Set the transaction time to be 20 minutes after the transaction date for one payment
+     */
+    payments[0].transaction.transaction_time = format(
+      getTime(
+        parse(
+          payments[0].transaction.transaction_date +
```
Comment (fw-noel) [inline apps/backend/test/unit/reconciliation/pos-reconciliation.spec.ts:44]: Can we use template literals \`${}\` here?

Also, an explanation for the 20 minutes thing? Is it just for the one test?

File: apps/backend/test/unit/reconciliation/pos-reconciliation.spec.ts [hunk 2]
```diff
@@ -1,58 +1,311 @@
...
+    expect(logger).toBeDefined();
+  });
+
+  it('should verify the test data transaction_time', () => {
+    expect(
+      differenceInMinutes(
+        parse(payments[0].transaction.transaction_time, 'HH:mm:ss', new Date()),
+        parse(deposits[0].transaction_time, 'HH:mm:ss', new Date())
+      )
+    ).toBeGreaterThanOrEqual(20);
```
Comment (fw-noel) [inline apps/backend/test/unit/reconciliation/pos-reconciliation.spec.ts:79]: What exactly are we testing here? Just that the 20 minute addition worked? It feels like we're just testing date-fns `differenceInMinutes`

File: apps/backend/test/unit/reconciliation/pos-reconciliation.spec.ts [hunk 3]
```diff
@@ -1,58 +1,311 @@
...
+    expect(deposits.every((itm) => itm.status === MatchStatus.PENDING)).toBe(
+      true
+    );
+  });
+
+  describe('matchPaymentsToDeposits heurisitics round one', () => {
+    it('should match payments to deposits according to first round of heuristics', () => {
+      const spy = jest.spyOn(service, 'matchPosPaymentToPosDeposits');
+      const roundOneTimeDiff = 5;
+      const timeDiff = roundOneTimeDiff;
```
Comment (fw-noel) [inline apps/backend/test/unit/reconciliation/pos-reconciliation.spec.ts:95]: Point of `roundOneTimeDiff`?

File: apps/backend/test/unit/reconciliation/pos-reconciliation.spec.ts [hunk 4]
```diff
@@ -1,58 +1,311 @@
...
+      /**
+       * Call function to match based on first round of heuristics
+       */
+      matchedRoundOne = service.matchPosPaymentToPosDeposits(
+        payments,
+        deposits,
+        timeDiff
+      );
+      matches = [...matchedRoundOne];
+      expect(spy).toBeCalled();
```
Comment (fw-noel) [inline apps/backend/test/unit/reconciliation/pos-reconciliation.spec.ts:106]: Is there a scenario that this test fails?

File: apps/backend/test/unit/reconciliation/pos-reconciliation.spec.ts [hunk 5]
```diff
@@ -1,58 +1,311 @@
...
+      /**
+       * Call function to match based on the second round of heuristics
+       */
+      matchedRoundTwo = service.matchPosPaymentToPosDeposits(
+        payments,
+        deposits,
+        timeDiff
+      );
+      matches.push(...matchedRoundTwo);
+      expect(spy).toBeCalled();
```
Comment (fw-noel) [inline apps/backend/test/unit/reconciliation/pos-reconciliation.spec.ts:148]: I'm still wondering if there's a scenario in which this fails. And if so, how this is different from the other `matchedRoundOne` test, since it's the same function. I might need to read more on how the heuristics are working

File: apps/backend/test/unit/reconciliation/pos-reconciliation.spec.ts [hunk 6]
```diff
@@ -1,58 +1,302 @@
...
+        ).toBeLessThanOrEqual(5)
+      );
+    });
+
+    it('should have updated the status from PENDING to MATCH', () => {
+      matchedRoundOne.forEach((itm) => {
+        expect(itm.deposit.status === MatchStatus.MATCH);
+        expect(itm.payment.status === MatchStatus.MATCH);
+      });
+    });
```
Comment (fw-noel) [inline apps/backend/test/unit/reconciliation/pos-reconciliation.spec.ts:113]: Do we know exactly how many matches we have, and how many non-matches we have? Is there an argument for adding that as a test too?

File: apps/backend/test/unit/reconciliation/helpers.ts [hunk 7]
```diff
@@ -0,0 +1,79 @@
+import { faker } from '@faker-js/faker';
+import { differenceInMinutes, format, getTime, parse } from 'date-fns';
+import { MatchStatus } from '../../../src/common/const';
+import { POSDepositEntity } from '../../../src/deposits/entities/pos-deposit.entity';
+import { PaymentEntity } from '../../../src/transaction/entities';
+import { paymentMethods } from '../../../test/mocks/const/payment-methods';
+import { locations } from '../../mocks/const/locations';
+import { MockData } from '../../mocks/mocks';
+
+export const setSomePaymentsToTwentyMinutesLater = (
```
Comment (fw-noel) [inline apps/backend/test/unit/reconciliation/helpers.ts:10]: This is also "resetting" the payments right (as in setting the status to pending)?

I'd suggest just passing an array into this function and calling it something like `resetPaymentsDataset` or something, and doing the 20 minute thing for each even index within the function rather than passing in an index. With some comments to explain why.

Comment (chelsea-EYDS) [reply apps/backend/test/unit/reconciliation/helpers.ts:9]: The payments are reset between tests using:
(this is a recent change :) )

```
  afterEach(() => {
    jest.clearAllMocks();
  });
```

Function has been updated to receive an array, and status updates are removed (it was _redundant_ since the mock data generates the status as PENDING and each test will generate new data)

:)

Comment (chelsea-EYDS) [reply apps/backend/test/unit/reconciliation/helpers.ts:9]: sorry for the caps-lock on PENDING, I wasn't yelling... that's how it is written in the codebase (as an enum)

File: apps/backend/invoke.sh [hunk 8]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
```
Comment (fw-noel) [inline apps/backend/invoke.sh:1]: Probably should be in a separate pr

File: apps/backend/test/unit/reconciliation/helpers.ts [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+            ...itm.transaction,
+            transaction_time: setTransactionTimeAheadBy20Minutes(
+              itm.transaction.transaction_date,
+              itm.transaction.transaction_time
+            )
+          }
+        }
+      : {
+          ...itm,
+          timestamp: itm.timestamp
```
Comment (fw-noel) [inline apps/backend/test/unit/reconciliation/helpers.ts:27]: Isn't `itm.timestamp` covered within `...itm`?

File: apps/backend/invoke.sh (hunk 8 above)
Comment (chelsea-EYDS) [reply apps/backend/invoke.sh:1]: oops wasn't planning to commit this one

**Review by fw-noel — APPROVED** [review]
